sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -389,6 +313,7 @@ object SimplifyBinaryComparison extends Rule[LogicalPlan] with PredicateHelper {
     case q: LogicalPlan => q transformExpressionsUp {
       // True with equality
       case a EqualNullSafe b if a.semanticEquals(b) => TrueLiteral
+      case a EqualNullSafe b if a.foldable || b.foldable => EqualTo(a, b)
```

Comment:
why do we need this? what if a or b are nullable?

Reply:
not needed any more